posthog/temporal/workflows/clickhouse.py

```diff
@@ -1,12 +1,273 @@
+        """
+        params = {**self.params}
+        if query_id is not None:
+            params["query_id"] = query_id
+
+        query = self.prepare_query(query, query_parameters)
+        request_data = self.prepare_request_data(data)
+
+        if request_data:
+            params["query"] = query
+        else:
+            request_data = query.encode("utf-8")
+
+        async with self.session.post(url=self.url, params=params, headers=self.headers, data=request_data) as response:
+            await self.check_response(response, query)
+            yield response
+
+    async def execute_query(self, query, *data, query_parameters=None, query_id: str | None = None) -> None:
+        """Execute the given query in ClickHouse.
+
+        This method doesn't return any response.
+        """
+        async with self.post_query(query, *data, query_parameters=query_parameters, query_id=query_id):
+            return None
+
+    async def read_query(self, query, *data, query_parameters=None, query_id: str | None = None) -> bytes:
+        """Execute the given query in ClickHouse and read the response in full.
+
+        As the entire payload will be read at once, use this method when expecting a small payload, like
+        when running a 'count(*)' query.
+        """
+        async with self.post_query(query, *data, query_parameters=query_parameters, query_id=query_id) as response:
+            return await response.content.read()
+
+    async def stream_query_as_jsonl(
+        self, query, *data, query_parameters=None, query_id: str | None = None, line_separator=b"\n"
+    ) -> typing.AsyncGenerator[dict[typing.Any, typing.Any], None]:
+        """Execute the given query in ClickHouse and stream back the response as one JSON per line.
+
+        This method makes sense when running with FORMAT JSONEachRow, although we currently do not enforce this.
+        """
+
+        buffer = b""
+        async with self.post_query(query, *data, query_parameters=query_parameters, query_id=query_id) as response:
+            async for chunk in response.content.iter_any():
+                lines = chunk.split(line_separator)
+
+                yield json.loads(buffer + lines[0])
+
+                buffer = lines.pop(-1)
+
+                for line in lines[1:]:
+                    yield json.loads(line)
+
+    async def stream_query_as_arrow(
+        self,
+        query,
+        *data,
+        query_parameters=None,
+        query_id: str | None = None,
+    ) -> typing.AsyncGenerator[dict[typing.Any, typing.Any], None]:
+        """Execute the given query in ClickHouse and stream back the response as Arrow record batches.
+
+        This method makes sense when running with FORMAT ArrowStreaming, although we currently do not enforce this.
+        As pyarrow doesn't support async/await buffers, we have to write the output to a local temporary file first.
+        """
+
+        with tempfile.SpooledTemporaryFile() as temp_file:
```

Comment:
We'll want to make the root path configurable through an envvar, so that we can point it into a k8s volume on prod. We don't have any guarantees about being allowed to write GB in the container's overlay.
The default on hobby and devenv and remain `/tmp`

Reply:
I've pushed a commit to get rid of the file and instead synchronously read ArrowStream batches. This is fairly memory friendly: during testing, batch memory usage was capped at memory required for 65409 rows (which indicates ClickHouse sets the batch size to be the [default block size](https://github.com/ClickHouse/ClickHouse/blob/f749b52f248f0811718d2b1b4fca7b243d8e2be6/src/Core/Defines.h#L22)).